File: ReactClient/src/Zones/ZoneManager.ts
import { toast } from "react-toastify";
import type { DangerZone, Zone } from "../Messages/AllTypes";

export class ZoneManager {
    private static instance: ZoneManager | null = null;
    private zoneIdToZone: Map<string, Zone>;
    private listeners: ((zones: Zone[]) => void)[] = [];

    private constructor() {
        this.zoneIdToZone = new Map<string, Zone>();
    }

    public static getInstance(): ZoneManager {
        if (this.instance === null) {
            this.instance = new ZoneManager();
        }
        return this.instance;
    }

    public subscribe(cb: (zones: Zone[]) => void): () => void {
        this.listeners.push(cb);
        return () => {
            this.listeners = this.listeners.filter(l => l !== cb);
        };
    }

    private notify() {
        const snapshot = this.getAllZones();
        this.listeners.forEach(cb => cb(snapshot));
    }

    tryAddZone(zone: Zone): boolean {
        if (!zone || !zone.zoneId) {
            toast.error("Invalid Zone or missing zoneId");
            return false;
        }

        if (this.zoneIdToZone.has(zone.zoneId)) {
            toast.error(`Zone with id ${zone.zoneId} already exists`);
            return false;
        }

        this.zoneIdToZone.set(zone.zoneId, zone);
        this.notify();
        return true;
    }

    tryRemoveZone(zoneId: string): boolean {
        if (!zoneId) return false;
        const removed = this.zoneIdToZone.delete(zoneId);
        if (removed) this.notify();
        return removed;
    }

    tryEditZone(zone: Zone): boolean {
        if (!zone || !zone.zoneId) {
            console.log("Invalid Zone or missing zoneId");
            return false;
        }

        if (!this.zoneIdToZone.has(zone.zoneId)) {
            console.log(`Zone with id ${zone.zoneId} does not exist`);
            return false;
        }

        this.zoneIdToZone.set(zone.zoneId, zone);
        this.notify();
        return true;
    }

    getZone(zoneId: string): Zone | undefined {
        if (!zoneId) return undefined;
        return this.zoneIdToZone.get(zoneId);
    }

    getAllZones(): Zone[] {
        return Array.from(this.zoneIdToZone.values());
    }

    getAllZoneIds(): string[] {
        return Array.from(this.zoneIdToZone.keys());
    }

    clearAll(): void {
        this.zoneIdToZone.clear();
        this.notify();
    }
}

File: C2TrainerServer/C2TrainerServer/Src/Zones/ZoneManager.cs
public class ZoneManager
{
    private static ZoneManager instance;
    private readonly Dictionary<string, Zone> _zones = new();

    private ZoneManager() { }

    public static ZoneManager GetInstance()
    {
        if (instance == null)
            instance = new ZoneManager();
        return instance;
    }

    public List<Zone> GetAllZones()
    {
        return _zones.Values.ToList();
    }

    public bool TryAddZone(Zone zone)
    {
        if (zone == null || string.IsNullOrWhiteSpace(zone.zoneId))
            return false;

        if (_zones.ContainsKey(zone.zoneId))
            return false;

        _zones[zone.zoneId] = zone;
        return true;
    }

    public Zone? TryGetZone(string zoneId)
    {
        if (string.IsNullOrWhiteSpace(zoneId))
            return null;

        _zones.TryGetValue(zoneId, out var zone);
        return zone;
    }

    public bool TryRemoveZone(string zoneId)
    {
        if (string.IsNullOrWhiteSpace(zoneId))
            return false;

        return _zones.Remove(zoneId);
    }

    public bool TryEditZone(string zoneId, Zone updatedZone)
    {
        if (string.IsNullOrWhiteSpace(zoneId) || updatedZone == null)
            return false;

        if (!_zones.ContainsKey(zoneId))
            return false;

        _zones[zoneId] = updatedZone;
        return true;
    }
}
